### PR TITLE
Fixed bugs with `rasa data convert`

### LINKED ISSUE #6505: Bugs with `rasa data convert`
Part of #6475

Fix bugs discovered during testing of `rasa data convert -f yaml`:

- [x] rasa data convert nlu: 
    1. [x] required args listed under optional args
    2. [x] Doesn’t recursively find files in the directory you provide (my data was in data/nlu/nlu.md)
    3. [x] Says it wants an output file, but requires an output directory 
- [x] NLU converter doesn't support external lookup tables
- [x] NLU converter doesn't ignore `.DS_store` file by default
- [x] NLU converter ignores `Retrieval Intents`
- [x] ignore `STORY_START` checkpoint
- [x] Why do we output GENR_OR checkpoints instead of `or:` statements? 

Originally from [here](https://docs.google.com/document/d/1LNzf4ZLefdOI-bFy4lzLTpvwT5CWbE9a2NZW45FtPCU/edit) and [here](https://www.notion.so/rasa/Feedback-on-2-0-4574eaacc0764967ac0b8b82fbc72904)

### PR DESCRIPTION
**Pull request by degiz**
Closes #6505 

**Proposed changes**:
- fixed docs
- not appending `STORY_START` checkpoint now
- properly convert `OR` statements
- `rasa data convert` now reads input folder recursively 
- added support for `retrieval intents` via `rasa data convert nlg -f yaml`
- added support for external `lookup tables`
- converter tests are now checking the content of output files

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/data.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def _find_data_files_in_directory(
     return filtered_files
 
 
-def _is_valid_filetype(path: Text) -> bool:
+def is_valid_filetype(path: Text) -> bool:
```
Comment (federicotdn) [inline rasa/data.py:176]: Let's add a docstring for this function.

File: rasa/nlu/training_data/formats/markdown_nlg.py [hunk 2]
```diff
@@ -81,6 +91,11 @@ def process_lines(lines: List[Text]) -> Dict[Text, List[Dict[Text, Text]]]:
 
         return responses
 
+    @staticmethod
+    def is_markdown_nlg_file(filename: Union[Text, Path]) -> bool:
```
Comment (federicotdn) [inline rasa/nlu/training_data/formats/markdown_nlg.py:95]: Docstring here as well!

File: rasa/cli/data.py [hunk 3]
```diff
@@ -254,57 +262,26 @@ def _convert_to_yaml(args: argparse.Namespace, is_nlu: bool) -> None:
...
-            f"Please note that in order for these stories to work you still "
-            f"need the 'FormPolicy' to be active. However the 'FormPolicy' is "
-            f"deprecated, please consider switching to the new 'RulePolicy', "
-            f"for which you can find the documentation here: {DOCS_URL_RULES}."
-        )
-
-    writer.dump(output_path, steps)
-
-    print_success(f"Converted Core file: '{source_path}' >> '{output_path}'.")
+    print_info(f"Converted {num_of_files_converted} file(s), saved in '{output}'.")
```
Comment (federicotdn) [inline rasa/cli/data.py:287]: Can we print another message if `num_of_files_converted` is 0?

**Review by federicotdn — COMMENTED** [review]
Thanks! I have no pressing comments to make on the code.

I tried this on [this](https://github.com/federicotdn/scry-bot/) bot, and I ran into some unexpected behaviour. The project's data structure is the following:
```
data/
├── core
│   ├── responses.md
│   └── stories.md
├── lookup_colors.txt
├── lookup_creatures.txt
└── nlu
    └── nlu.md

2 directories, 5 files
```
I created a new directory called `new/`, to use as output directory for the converter. The first command I tried was:
```bash
rasa data convert nlg --data data/core/responses.md --out new -f yaml
```
The output was:
```bash
Converted 0 file(s), saved in 'new'.
```
Here I was expecting `responses.md` to be detected. Is this correct?

After that, I tried again but specifying the entire `data/` directory instead:
```bash
rasa data convert nlg --data data --out new -f yaml
```
The output was:
```bash
Converted NLG file: 'data/core/responses.md' >> 'new/responses_converted.yml'.
Converted Core file: 'data/core/stories.md' >> 'new/stories_converted.yml'.
Converted 2 file(s), saved in 'new'.
```
So here the `responses.md` was correctly detected, but for some reason `stories.md` was picked up as well. I assume this shouldn't have happened (I think the `if-elif-else` chain in `_convert_to_yaml` is the cause).

File: rasa/cli/data.py (hunk 3 above)
Comment (degiz) [reply rasa/cli/data.py:287]: good point!

**Comment by degiz** [comment]
@federicotdn thanks for finding this bug!

I've introduced a new interface `TrainingDataConverter`, and it's implementations are now taking care of NLU, Stories, NLG separately.

File: rasa/cli/data.py [hunk 4]
```diff
@@ -254,57 +281,23 @@ def _convert_to_yaml(args: argparse.Namespace, is_nlu: bool) -> None:
...
+    else:
         print_warning(
-            f"Training data file '{source_path}' contains forms. "
-            f"Any 'form' events will be converted to '{KEY_ACTIVE_LOOP}' events. "
-            f"Please note that in order for these stories to work you still "
-            f"need the 'FormPolicy' to be active. However the 'FormPolicy' is "
-            f"deprecated, please consider switching to the new 'RulePolicy', "
-            f"for which you can find the documentation here: {DOCS_URL_RULES}."
+            f"Didn't convert any file in '{training_data}' folder. "
+            "Did you specify the correct folder?"
```
Comment (federicotdn) [inline rasa/cli/data.py:302]:
```suggestion
            f"Didn't convert any files in '{training_data}' directory. "
            "Did you specify the correct directory?"
```

File: rasa/utils/tensorflow/converter.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from pathlib import Path
```
Comment (federicotdn) [inline rasa/utils/tensorflow/converter.py:1]: Should this file really be in `rasa.utils.tensorflow`? Does it have any specific relation with TF?

File: rasa/core/training/converters/story_markdown_to_yaml_converter.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        writer.dump(output_core_path, steps)
+
+        print_success(f"Converted Core file: '{source_path}' >> '{output_core_path}'.")
+
+    @classmethod
+    def converted_file_suffix(cls) -> Text:
+        """Returns suffix that should be appended to the converted training data file.
+        """
+        return "_converted.yml"
```
Comment (federicotdn) [inline rasa/core/training/converters/story_markdown_to_yaml_converter.py:64]: Looks like all three implementation of this method are identical. We could implement it directly in `TrainingDataConverter`.

**Review by federicotdn — COMMENTED** [review]
Ok! Now Core files are not being picked up anymore :+1:

However, I tested specifying a file with `--data` (`rasa data convert nlg --data data/core/responses.md --out new -f yaml`) and it still didn't work. Since the `--help` output shows that `--data` accepts files and directories (`Path to the file or directory containing Rasa data. (default: data)`, I still think we should fix this. Core and NLU should also accept a single file to convert.

File: rasa/utils/tensorflow/converter.py (hunk 5 above)
Comment (degiz) [reply rasa/utils/tensorflow/converter.py:1]: Oh my god, definitely no! Supposed to be one level up, thanks for spotting that!

File: rasa/core/training/converters/story_markdown_to_yaml_converter.py (hunk 6 above)
Comment (degiz) [reply rasa/core/training/converters/story_markdown_to_yaml_converter.py:64]: The idea was that `TrainingDataConverter` can be used for `any format` -> `any format` conversion, but you're probably right. This can be changed if the future if necessary.

**Review by federicotdn — APPROVED** [review]
Thanks!! :+1:

**Comment by akelad** [comment]
@degiz i'm in the process of reviewing the CLI docs generally speaking, but since you're making changes to `rasa data convert`, can you make sure that section is up to date? doesn't look like it is right now: https://rasa.com/docs/rasa/next/command-line-interface#rasa-data-convert

**Comment by degiz** [comment]
@akelad I think I'm already changing that with this PR 👍

**Comment by akelad** [comment]
you aren't, you've only updated the table at the top of the page, the description further down still only refers to `rasa data convert nlu`

**Comment by akelad** [comment]
also i'm not sure each `rasa data convert nlu` `rasa data convert core` should be listed in the table, but rather only `rasa data convert`, and the rest can be described in the section further down?

**Comment by akelad** [comment]
can we remove the `ready-to-merge` label until that's done please?

File: docs/docs/migration-guide.mdx [hunk 7]
```diff
@@ -37,6 +37,7 @@ to the new `YAML` format using following commands:
 
 ```bash
 rasa data convert nlu -f yaml --data={SOURCE_DIR} --out={TARGET_DIR}
+rasa data convert nlg -f yaml --data={SOURCE_DIR} --out={TARGET_DIR}
```
Comment (akelad) [inline docs/docs/migration-guide.mdx:40]: i guess if we're adding this, then above you should also say that you can convert responses.md with that command?

File: docs/docs/command-line-interface.mdx [hunk 8]
```diff
@@ -279,11 +279,31 @@ You can start the converter by running:
...
+
+You can specify the input file/directory, output directory with the following arguments:
+
+```bash
+rasa data convert core --help
+```
+
+## rasa data convert nlg
+
+You can convert NLG data from Markdown to YAML.
```
Comment (akelad) [inline docs/docs/command-line-interface.mdx:299]: what is NLG data? Do you mean the responses.md for the response selector?

Comment (akelad) [reply docs/docs/command-line-interface.mdx:299]: also can you add `nlg` to the word exclusion list for typo CI please

File: docs/docs/command-line-interface.mdx [hunk 9]
```diff
@@ -279,11 +279,31 @@ You can start the converter by running:
 rasa data convert nlu
 ```
 
-You can specify the input file, output file, and the output format with the following arguments:
+You can specify the input file/directory, output file/directory, and the output format with the following arguments:
```
Comment (akelad) [inline docs/docs/command-line-interface.mdx:282]:
```suggestion
You can specify the input file or directory, output file or directory, and the output format with the following arguments:
```

File: docs/docs/command-line-interface.mdx [hunk 10]
```diff
@@ -279,11 +279,31 @@ You can start the converter by running:
...
 
 ```text [rasa data convert nlu --help]
 ```
 
+## rasa data convert core
+
+You can convert Core data from Markdown to YAML.
+
+You can specify the input file/directory, output directory with the following arguments:
```
Comment (akelad) [inline docs/docs/command-line-interface.mdx:291]:
```suggestion
You can specify the input file or directory and the output directory with the following arguments:
```

File: docs/docs/command-line-interface.mdx [hunk 11]
```diff
@@ -279,11 +279,31 @@ You can start the converter by running:
...
+
+```bash
+rasa data convert core --help
+```
+
+## rasa data convert nlg
+
+You can convert NLG data from Markdown to YAML.
+
+You can specify the input file/directory, output directory with the following arguments:
```
Comment (akelad) [inline docs/docs/command-line-interface.mdx:301]:
```suggestion
You can specify the input file or directory and output directory with the following arguments:
```

**Review by akelad — COMMENTED** [review]
thanks for adding, just a few changes to be made!

File: docs/docs/command-line-interface.mdx (hunk 8 above)
Comment (degiz) [reply docs/docs/command-line-interface.mdx:299]: > what is NLG data? Do you mean the responses.md for the response selector?
Yes! mentioned that in the docs above

> also can you add nlg to the word exclusion list for typo CI please
✅

Comment (akelad) [reply docs/docs/command-line-interface.mdx:299]: hmmm, i would still explicitly mention that here. and did we officially decide we're calling responses.md NLG now? i can also start that discussion on slack though, i'm gonna be re-reviewing this page later anyways

Comment (degiz) [reply docs/docs/command-line-interface.mdx:299]: What I've added is:

> You can convert existing NLU, Stories, and NLG (i.e. `responses.md`) training data

Let's indeed discuss that on Slack, I'm happy to rename NLG to whatever we decide on. Currently NLG is a term we use in the code itself.

**Comment by degiz** [comment]
@wochinge This change doesn't put anything new to the `shared`, just modifying already existing files.

**Comment by wochinge** [comment]
Will we need this in Rasa X @degiz ?

**Comment by degiz** [comment]
Currently I don't think so, we don't plan to convert formats.

**Comment by wochinge** [comment]
Amazing, thanks 🙌
